File: leetcode_problems/typescript/src/lc3090_max_len_substring_with_two_occur.ts
function maximumLengthSubstring(s: string): number {
  let map: Map<string, number> = new Map();
  map.set(s[0], 1);
  map.set(s[1], (map.get(s[1]) || 0) + 1);
  let max = 2;
  let p1 = 0;
  for (let p2 = 2; p2 < s.length; p2++) {
    map.set(s[p2], (map.get(s[p2]) || 0) + 1);
    while (p1 < s.length && (map.get(s[p2]) || 0) > 2) {
      map.set(s[p1], (map.get(s[p1]) || 0) - 1);
      p1++;
    }
    max = Math.max(max, p2 - p1 + 1);
  }
  return max;
}

File: leetcode_problems/cpp/lc3090_max_len_substring_with_two_occur.cpp
#include <string>
#include <algorithm>
using std::string;

class Solution
{
public:
    int maximumLengthSubstring(string s)
    {
        int map[26];
        map[s[0] - 'a']++;
        map[s[1] - 'a']++;
        int max = 2;
        int p1 = 0;
        for (int p2 = 2; p2 < s.size(); p2++)
        {
            map[s[p2] - 'a']++;
            while (p1 < s.size() && map[s[p2] - 'a'] > 2)
            {
                map[s[p1] - 'a']--;
                p1++;
            }
            max = std::max(max, p2 - p1 + 1);
        }
        return max;
    }
};
